### PR TITLE
Add instrumentation and retry configurations

### PR DESCRIPTION
**Pull request by xichen2020**
cc @jeromefroe @martin-mao @cw9  @robskillington 

This PR adds metrics configurations to `instrument/` for reuse across repos.

### CODE REVIEW

File: instrument/config.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
Comment (cw9) [inline instrument/config.go]: nit: 2017?

File: instrument/config.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+// SampleRate returns the metrics sampling rate.
+func (mc *MetricsConfiguration) SampleRate() float64 {
+	if mc.SamplingRate > 0.0 && mc.SamplingRate <= 1.0 {
+		return mc.SamplingRate
+	}
+	return defaultSamplingRate
+}
+
+// ReportInterval returns the metrics reporting interval.
+func (mc *MetricsConfiguration) ReportInterval() time.Duration {
```
Comment (cw9) [inline instrument/config.go]: does this need to be exported?

File: instrument/config.go (hunk 1 above)
Comment (xichen2020) [inline instrument/config.go]: Yeah good catch, will fix.

File: instrument/config.go (hunk 2 above)
Comment (xichen2020) [inline instrument/config.go]: Yeah the instrumentation options have a `ReportInterval` property that's used to control how frequently gauge metrics are reported in separate goroutines (e.g., number of file descriptors, and all the database related gauge metrics in m3db).

**Pull request merged**